server/controllers/auth/signup.js

```diff
@@ -1,3 +1,39 @@
-exports.signup = (req, res) => {
-  res.send(' Hello from signup file');
+const { hash } = require('bcrypt');
+const { sign } = require('jsonwebtoken');
+const { signupSchema } = require('./validation/signup-validation');
+const { users: { getUserByUsername, getUserByEmail, addUser } } = require('../../database/queries');
+
+
+exports.signup = (req, res, next) => {
+  const key = process.env.KEY;
+  const {
+    username, email, password, nativeLangId, learningLangId,
+  } = req.body;
+
+  signupSchema.validate({
+    username, email, password, nativeLangId, learningLangId,
+  })
+    .then(() => getUserByUsername(username))
+    .then((result) => {
+      if (result.rows.length !== 0) throw Error('username exists');
+    })
+    .then(() => getUserByEmail(username))
+    .then(({ rows }) => {
+      if (rows.length !== 0) throw Error('email exists');
+    })
+    .then(() => hash(password, 10))
+    .then((hashed) => addUser({
+      username, email, password: hashed, nativeLangId, learningLangId,
+    }))
```

Comment:
```
signupSchema.validate({
    username, email, password, nativeLangId, learningLangId,
})
    .then(() => Promise.all([
        getUserByUsername(username),
        getUserByEmail(username),
        hash(password, 10),
    ]))
    .then(([usersByUsername, usersByEmail, pwdHash]) => {
        if (usersByUsername.rows.length !== 0) throw Error('username exists')
        else if (usersByEmail.length !== 0) throw Error('email exists');

        return addUser({
            username, email, password: pwdHash, nativeLangId, learningLangId,
        });
    });
```
This will execute both of the database requests and the password hashing simultaneously, instead of doing it sequentially one after the other. (On the other hand it will always execute all three actions, even when for example the username already exists - but that's a tradeoff worth doing IMO)

Reply:
and actually, you can (and should) create another query: `getUserByUsernameAndEmail()`, so that you can do only one roundtrip to the database to check if the username or email exists, instead of two..

Reply:
good note thank you